### DISCUSSION TITLE
Make it easier to build and run tests

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#9857] [Ideas]
At the moment it's a bit of a PITA to build Keycloak, especially following the instructions in https://github.com/keycloak/keycloak/blob/main/docs/building.md.

I'm also having constant issues with InteliJ and weird build errors there. Not sure if that's a common thing, but seems all the `magic` stuff around the new store, javascript adapters, npm modules, etc. are all a bit of a PITA?

Some issues with it:

* `mvn clean install` builds and runs the testsuite, which takes an enormous amount of time
* Asking people to add nexus to their POMs is just annoying like hell

There's also some issues with missing dependencies especially around the testsuite not having dependencies on the distribution.

A couple things here that will help:

* Getting rid of WildFly distribution will allow us to clean up a lot
* Getting rid of old examples
* Making the testsuite something that has to be explicitly built/execute rather than part of the `default` build
* Cleanup of modules, and making sure there are dependencies between modules when needed
* Splitting the server and adapters into separate things

**Comment by stianst** [comment]
I'd propose that:

    mvn install

Would not include the testsuite at all, then you could choose to run everything with:

    mvn install -Ptestsuite

Or, just run the testsuite with:

    mvn -f testsuite install 

Further, when the testsuite is enabled from the root, the distribution should also be built, and we need to make sure there are dependencides on distribution bits within testsuite so a Maven reactor build works properly. For example the following should work:

    mvn install -Ptestsuite -pl testsuite -am

Or, if I want to rebuild the server dist, then run testsuite I could do:

    mvn -Ddistribution -pl distribution/server install 
    mvn -f testsuite install

**Reply by stianst** [reply]
Basically using `pl` and `am` should works as expected

**Comment by kresnasatya** [comment]
I really like this idea! Today, I just fork Keycloak repo and run `mvn clean install` then I got error build failure as I make this discussion: [Cannot Build Keycloak from GitHub Repo](https://github.com/keycloak/keycloak/discussions/11997) and looking forward for the solution.

**Reply by kresnasatya** [reply]
It's solved!

**Comment by cgeorgilakis** [comment]
This is great news. Until version 16.1.0, everything is working well with run/debug/ run tests from Intellij with rare errors.
After migrating to keycloak version 18 ( maybe from 17), I have problems starting IDELauncher in Intelli (community edition). I could run tests if I select 'Do not to build before running'.
However mvn commands are working, this is time consuming process.

As @stianst said attention must be in Keycloak X ( quarkus).

Moreover, I believe that you should try to migrate all main Keycloak modules to Java 11. It is problematic using version Java 7 (!) in some modules and use simultaneously Quarkus which is running in Java 11 at least. This make also problems in IDE. I could try make a PR for this if Keycloak team will be intrested.
